### PR TITLE
persist: PackedNumeric encoding

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR adds a `PackedNumeric` encoding whose byte value sorts the same as the Rust `Numeric` struct.

This implementation is **different than what was called out in the design doc**. Originally we planned to encode two values, a `(float, <opaque bytes>)` tuple most likely in an Arrow struct column. This is what I originally implemented but extensive benchmarking showed this to be less performant than encoding a bit-swizzled version of the original bytes from the `Numeric`, which ultimately is what this PR implements.

#### Benchmarking

Here is a [gist](https://gist.github.com/ParkMyCar/9f78197659ea02cc567ad82028e65e63) of the alternative implementation. For benchmarking I spun up the same EC2 instance we run in production and ran everything there.

The benchmark itself was quite simple, either encode or decode 1,000,000,000 times, and measure with `perf`, e.g.

```
for _ in 0..1_000_000_000 {
    let packed = std::hint::black_box(PACKED).into_value();
    std::hint::black_box(packed);
}
```

What I found is for _decoding_ the bit-swizzling implementation is ~15% slower than the "plain" implementation, even though it has a lot more logic. What happens is the CPU is able to concurrently run the bit shift and arithmetic operations with memory accesses, `perf` validates this by looking at the number of instructions run per-cycle:

impl | instructions per cycle |
-----|------------------------|
bit-swizzling | 1.04
plain         | 0.63

For _encoding_ the bit-swizzling implementation is ~20% _faster_ than the "plain" implementation, even without the additional step to also encode a float. Again let's compare the instructions per-cycle:

impl | instructions per cycle |
-----|------------------------|
bit-swizzling | 2.09
plain         | 0.87

With the bit-swizzling implementation it appears as though the CPU is able to order things more efficiently and spends less time waiting on reads from memory. For the encoding step when you also generate a `float32` along with the "plain" encoded version, the bit-swizzling impl is 3x faster. I believe the culprit here is converting a numeric to `float32` requires calling into a C++ library, which introduces overhead. It's technically possible to get C++ function calls inlined into Rust, but we're pretty far away from being able to do that.

So even though decoding the bit-swizzled implementation is a bit slower, I opted for that implementation because the encoding speed is much faster. More than happy to merge the "plain" encoding if that's what folks would prefer.

### Motivation

Progress towards https://github.com/MaterializeInc/materialize/issues/24830

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a

### CODE REVIEW

File: src/repr/src/adt/numeric.rs [hunk 1]
```diff
@@ -762,11 +763,111 @@ impl DecimalLike for Numeric {
...
+        let (digits, exponent, mut bits, lsu) = value.to_raw_parts();
+
+        let mut buf = [0u8; 40];
+
+        // XORing the exponent will ensure negative exponents (smaller) sort before positive
+        // exponents (larger).
+        let exponent = (exponent as u32) ^ 0x8000_0000u32;
+
+        // Naturally the marker bits are in reverse sorted order, except for all positive numbers.
+        // In this case adding 96 and then inverting the bits provides the right sort.
```
Comment (bkirwi) [inline src/repr/src/adt/numeric.rs:805]: Not actually adding!

File: src/repr/src/adt/numeric.rs [hunk 2]
```diff
@@ -762,11 +763,111 @@ impl DecimalLike for Numeric {
...
+        if bits == 0 {
+            bits = 96;
+        }
+        let marker = (!bits) as u32;
```
Comment (bkirwi) [inline src/repr/src/adt/numeric.rs:809]: OOC, why u32 if it fits in a byte?

File: src/repr/src/adt/numeric.rs [hunk 3]
```diff
@@ -784,4 +885,43 @@ mod tests {
             assert_eq!(actual.unwrap(), expect);
         }
     }
+
+    #[mz_ore::test]
+    fn proptest_packed_numeric_sorts() {
```
Comment (bkirwi) [inline src/repr/src/adt/numeric.rs:890]: 🔥

**Review by bkirwi — COMMENTED** [review]
This is pretty clearly working and very cool, but the bit-swizzling of the internal representation makes me nervous... particularly because `Numeric` documents those internals as unspecified and subject to change , and we're trying to choose a serialization that will be stable over time.

We could also consider:
- Explicitly documenting what the bits of the serialized representation mean.
- Basing the serialization on the BCD output, since that's more explicitly documented and may be a more stable target.

Anyways what do you think?

File: src/repr/src/adt/numeric.rs (hunk 1 above)
Comment (ParkMyCar) [reply src/repr/src/adt/numeric.rs:805]: Ahh shoot, yeah let me update the comment

File: src/repr/src/adt/numeric.rs (hunk 2 above)
Comment (ParkMyCar) [reply src/repr/src/adt/numeric.rs:809]: We get better performance when writing this out as a `u32`, ARM doesn't seem to like writing out single bytes

**Comment by ParkMyCar** [comment]
> This is pretty clearly working and very cool, but the bit-swizzling of the internal representation makes me nervous... particularly because `Numeric` documents those internals as unspecified and subject to change , and we're trying to choose a serialization that will be stable over time.

Totally fair! Using the internal representation also makes me a bit nervous, FWIW I don't think it will ever change, and we own the bindings to the underlying C++ library, but it totally could happen.

> Basing the serialization on the BCD output, since that's more explicitly documented and may be a more stable target.

I like this idea, my only concern is performance. Let me whip this up and see how it compares!

**Comment by ParkMyCar** [comment]
Oof, a quick first attempt at using BCD is ~4x slower than this version. I largely suspect it's because `to_packed_bcd(...)` [allocates](https://github.com/MaterializeInc/rust-dec/blob/master/dec/src/decimal.rs#L575) which we could avoid. Will think on this

**Comment by bkirwi** [comment]
Ah, that's annoying! If you otherwise like BCD we could always put in the less efficient impl and come back to it... since we store this in proto as BCD that should be no less efficient than today, and it looks like the allocation is in the bindings we control.

But anyways I'm certainly not stuck on BDC... that's just the first representation I saw that a) looks like it would be possible to make a sortable representation from, and b) has a documented format I can understand.

**Comment by ParkMyCar** [comment]
@benesch here's the Numerics PR I mentioned in the Persist Sync today.

tl;dr the original proposal was storing `(float, opaque bytes)` but calling into `libdecimal` to convert to a floating point representation is relatively slow. The alternative I implemented was a bit swizzled version of the internal repr of `libdecimal`, which sorts the same as the Rust type and is much faster for encoding (no need to call into `libdecimal`) and just slightly slower for decoding.

The question we're going back and forth on is "are we okay durably persisting bytes based on the internal representation of `libdecimal`?".

**Comment by benesch** [comment]
> * Explicitly documenting what the bits of the serialized representation mean.

I'd be up for taking a PR to do this in rust-dec. The underlying libdecnumber library will likely never change again.

Here's libdecnumber's "documentation" of what the fields mean: https://github.com/gcc-mirror/gcc/blob/2b438a0d2aa80f051a09b245a58f643540d4004b/libdecnumber/decNumber.h#L76-L102

**Comment by ParkMyCar** [comment]
Eek! turns out there is a subtle bug in the bit swizzle impl that a better proptest strategy revealed. I think it's fixable but I've also put too much time into this already. I'm just going to switch to the simpler implementation of `(float32, <opaque bytes>)`

File: src/repr/src/adt/numeric.rs [hunk 4]
```diff
@@ -762,11 +763,161 @@ impl DecimalLike for Numeric {
...
+/// * **num_digits** is encoded as big endian.
+///
+/// ## Notes:
+///
+/// * Even though the encoded format only requires 38 bytes, we allocate 40 for
+///   better cache alignment, since 40 is a multiple of 8.
+///
+/// [1]: <https://github.com/gcc-mirror/gcc/blob/2b438a0d2aa80f051a09b245a58f643540d4004b/libdecnumber/decNumber.h#L76-L102>
+/// [2]: <http://speleotrove.com/decimal/dnnumb.html>
+/// [`OrderedDecimal`]: dec::OrderedDecimal
```
Comment (bkirwi) [inline src/repr/src/adt/numeric.rs:819]: Thanks - this is extremely helpful!

Comment (ParkMyCar) [reply src/repr/src/adt/numeric.rs]: No problem! Just a heads up though, I found a subtle bug in this impl and decided to ditch it 😅

Comment (bkirwi) [reply src/repr/src/adt/numeric.rs]: 😅

File: src/repr/src/adt/numeric.rs [hunk 5]
```diff
@@ -762,11 +763,76 @@ impl DecimalLike for Numeric {
...
+    fn from_bytes(slice: &[u8]) -> Result<Self, String> {
+        let buf: [u8; Self::SIZE] = slice.try_into().map_err(|_| {
+            format!(
+                "size for PackedNumeric is {} bytes, got {}",
+                Self::SIZE,
+                slice.len()
+            )
+        })?;
+        Ok(PackedNumeric(buf))
+    }
```
Comment (def-) [inline src/repr/src/adt/numeric.rs:789]: Untested

File: src/repr/src/adt/numeric.rs [hunk 6]
```diff
@@ -762,11 +763,76 @@ impl DecimalLike for Numeric {
...
+/// [`PackedNumeric`] sorts the same as [`Numeric`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct PackedNumeric(pub [u8; 40]);
+
+impl FixedSizeCodec<Numeric> for PackedNumeric {
+    const SIZE: usize = 40;
+
+    fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
```
Comment (def-) [inline src/repr/src/adt/numeric.rs:778]: (Untested outside of unit tests)

File: src/repr/src/adt/numeric.rs [hunk 7]
```diff
@@ -762,11 +763,76 @@ impl DecimalLike for Numeric {
...
+            let x: [u8; 2] = self.0[(i * 2) + 8..(i * 2) + 10].try_into().unwrap();
+            let x = u16::from_le_bytes(x);
+            lsu[i] = x;
+        }
+
+        let bits: [u8; 1] = self.0[34..35].try_into().unwrap();
+        let bits = u8::from_le_bytes(bits);
+
+        Numeric::from_raw_parts(digits, exponent, bits, lsu)
+    }
```
Comment (def-) [inline src/repr/src/adt/numeric.rs:826]: Only tested in unit tests and looks very complex! Might be a good candidate for SLT/TD

**Review by def- — CHANGES_REQUESTED** [review]
Nightly shows a Miri test failure: https://buildkite.com/materialize/nightly/builds/8077 (the rest is fine) I'll push a change to disable these tests in Miri
```
error: unsupported operation: can't call foreign function `decNumberCopyNegate` on OS `linux`
```
Coverage report: https://buildkite.com/materialize/coverage/builds/426#_ I marked some interesting spots inline, is it possible to add explicit integration tests for them or is a difficult setup required? If the latter, I can help out.

File: src/repr/src/adt/numeric.rs (hunk 5 above)
Comment (ParkMyCar) [reply src/repr/src/adt/numeric.rs:789]: Added a test to cover this!

File: src/repr/src/adt/numeric.rs (hunk 7 above)
Comment (ParkMyCar) [reply src/repr/src/adt/numeric.rs:826]: Only being covered in unit tests is expected! This encoding will be used for the V1 of columnar encodings which isn't fully hooked up yet

File: src/repr/src/row/encoding2.rs [hunk 8]
```diff
@@ -183,10 +190,25 @@ impl DatumColumnEncoder {
...
-                builder.append_value(&bytes);
+            (
+                DatumColumnEncoder::Numeric {
+                    approx_values,
+                    binary_values,
+                },
+                Datum::Numeric(val),
+            ) => {
+                let mut ctx = Context::default();
+                let float_approx = ctx.try_into_f64(val.0).unwrap_or_else(|_| {
```
Comment (bkirwi) [inline src/repr/src/row/encoding2.rs:201]: Oh, this is interesting! I take it that the only case where converting to f64 fails is when the numeric is out of range?

File: src/repr/src/row/encoding2.rs [hunk 9]
```diff
@@ -480,7 +508,29 @@ impl DatumColumnEncoder {
...
-            DatumColumnEncoder::Numeric(mut builder) => Arc::new(builder.finish()),
+            DatumColumnEncoder::Numeric {
+                mut approx_values,
+                mut binary_values,
+            } => {
+                let approx_array = approx_values.finish();
+                let binary_array = binary_values.finish();
+
+                assert_eq!(approx_array.len(), binary_array.len());
+                assert_eq!(approx_array.logical_nulls(), binary_array.logical_nulls());
```
Comment (bkirwi) [inline src/repr/src/row/encoding2.rs:519]: This looks like it's `O(n)`... maybe this should be a debug assert?

**Review by bkirwi — APPROVED** [review]
Nice; thanks for the followup!

File: src/repr/src/row/encoding2.rs (hunk 8 above)
Comment (ParkMyCar) [reply src/repr/src/row/encoding2.rs:201]: Exactly! Reading the code I don't think this should ever happen, but adding this just to be safe

File: src/repr/src/row/encoding2.rs (hunk 9 above)
Comment (ParkMyCar) [reply src/repr/src/row/encoding2.rs]: Good call! Updated!
